opentracing/recorder.go

```diff
@@ -0,0 +1,90 @@
+package opentracing
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	basictracer "github.com/opentracing/basictracer-go"
+	"sourcegraph.com/sourcegraph/appdash"
+)
+
+// Recorder implements basictracer.Recorder.
+type Recorder struct {
+	collector appdash.Collector
+	logOnce   sync.Once
+	verbose   bool
+}
+
+// NewRecorder forwards basictracer.RawSpans to an appdash.Collector.
+func NewRecorder(collector appdash.Collector, opts Options) *Recorder {
+	return &Recorder{
+		collector: collector,
+		verbose:   opts.Verbose,
+	}
+}
+
+// RecordSpan converts a RawSpan into the Appdash representation of a span
+// and records it to the underlying collector.
+func (r *Recorder) RecordSpan(sp basictracer.RawSpan) {
+	spanID := appdash.SpanID{
+		Span:   appdash.ID(uint64(sp.SpanID)),
+		Trace:  appdash.ID(uint64(sp.TraceID)),
+		Parent: appdash.ID(uint64(sp.ParentSpanID)),
+	}
+
+	r.collectEvent(spanID, appdash.SpanName(sp.Operation))
+
+	// Record all of the logs. Payloads are thrown out.
+	for _, log := range sp.Logs {
+		r.collectEvent(spanID, appdash.LogWithTimestamp(log.Event, log.Timestamp))
+	}
+
+	for key, value := range sp.Tags {
+		val := []byte(fmt.Sprintf("%v", value))
+		r.collectAnnotation(spanID, appdash.Annotation{Key: key, Value: val})
+	}
+
+	for key, val := range sp.Baggage {
+		r.collectAnnotation(spanID, appdash.Annotation{Key: key, Value: []byte(val)})
```

Comment:
I might be totally misreading that, but it sounds like it means that baggage must be propagated out-of-process to downstream processes? e.g. baggage items need to be transferred over the network, into another backend process that would be running appdash as well (appdash in a distributed context), whereas tags are directly communicated to the storage backend for later retrieval by someone who is viewing the trace, yeah?

If so, I think this means that we would need some sort of special handling for baggage items. By default, annotations that are collected by an `appdash.Collector` are not inherently propagated in any special way. In fact, they are not strictly transferred over the network at all (because appdash can, and does, run in-process). The only thing that `appdash.Collector` guarantees is that, once passed off for collection, spans will be available / presentable to the user (read: developer looking at traces) later on via the web UI for example.

How are the other tracing system implementations propagating these downstream? I wonder if it's left up to the user to attach these as metadata with e.g. gRPC requests or as headers on HTTP requests, or if they do something else?